sns_ik_lib/include/sns_ik/sns_vel_ik_base.hpp

```diff
@@ -0,0 +1,191 @@
+
+class SnsVelIkBase {
+
+public:
+
+  // Smart pointer typedefs. Note: all derived classes MUST override these smart pointers.
+  typedef std::shared_ptr<SnsVelIkBase> Ptr;
+  typedef std::unique_ptr<SnsVelIkBase> uPtr;
+
+  enum class ExitCode {
+    Success,  // successfully solved the optimization problem
+    BadUserInput,  // user input failed basic checks (eg. inconsistent matrix size)
+    InfeasibleTask,  // there is no feasible solution to the primary task
+    InternalError  // there was an internal error in the solver (should never happen...)
+  };
+
+  /**
+   * Create a default solver with nJnt joints and no bounds on joint velocity
+   * @param nJnt: number of joints in the robot model (columns in the jacobian)
+   * @return: velocity solver iff successful, nullptr otherwise
+   */
+  static std::unique_ptr<SnsVelIkBase> create(int nJnt);
+
+  /**
+   * Create a default solver with constant bounds on the joint velocity
+   * @param dqLow: lower bound on the velocity of each joint
+   * @param dqUpp: upper bound on the velocity of each joint
+   * @return: velocity solver iff successful, nullptr otherwise
+   */
+  static std::unique_ptr<SnsVelIkBase> create(const Eigen::ArrayXd& dqLow,
+                                              const Eigen::ArrayXd& dqUpp);
+
+  // Make sure that class is cleaned-up correctly
+  virtual ~SnsVelIkBase() {};
+
+  /**
+   * Set the bounds on joint velocity.
+   * Set a bound to infinity (or an arbitrarily large value) to disable it.
+   * This method will update the number of joints in the solver
+   * Requirements: inputs must be the same size and dqLow < dqUpp
+   * @param dqLow: lower bound on the velocity of each joint
+   * @param dqUpp: upper bound on the velocity of each joint
+   * @return: true iff successful
+   */
+  bool setVelBnd(const Eigen::ArrayXd& dqLow, const Eigen::ArrayXd& dqUpp);
+
+  /**
+   * Solve a velocity IK problem with no null-space bias of joint-space optimization.
+   *
+   *  This method implements   "Algorithm 1: SNS algorithm"   from the paper:
+   *  "Control of Redundant Robots Under Hard Joint Constraint: Saturation in the Null Space"
+   *   by: Fabrizio Flacco, Alessandro De Luca, Oussama Khatib
+   *
+   * Solve for joint velocity dq and task scale s:
+   *
+   *  maximize: s
+   *  subject to:
+   *    s * dx = J * dq
+   *    0 < s <= 1
+   *    dqLow <= dq <= dqUpp      ( bounds set in constructor or setVelBnd() )
+   *
+   * @param J: Jacobian matrix, mapping from joint to task space. Size = [nTask, nJoint]
+   * @param dx: task velocity vector. Length = nTask
+   * @param[out] dq: joint velocity solution. Length = nJoint
+   * @param[out] taskScale: task scale.  fwdKin(dq) = taskScale*dx
+   *                            taskScale == 1.0  --> task was feasible
+   *                            taskScale < 1.0  --> task was infeasible and had to be scaled
+   * @return: Success: the algorithm worked correctly and satisfied the problem statement
```

Comment:
It returns an ExitCode technically

Reply:
Good catch - I will fix it in both places.